File: lab2/c/barrier.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void computations() {
        double x = 123456789;
        for (int i = 0; i < 100; i++)
	        x = tan(exp(log(x)));
}

int main(int argc, char** argv) {
	MPI_Init(&argc, &argv);
	int world_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
	int world_size;
	MPI_Comm_size(MPI_COMM_WORLD, &world_size);

	if (world_size < 2) {
		fprintf(stderr, "World size must be greater than 1 for %s\n", argv[0]);
		MPI_Abort(MPI_COMM_WORLD, 1); 
	}

	int iterations = 10000;

	double start = MPI_Wtime();
	for (int i = 0; i < iterations; i++) {
		MPI_Barrier(MPI_COMM_WORLD);
		computations();
	}
	double end = MPI_Wtime();
	double elapsed_barrier = end - start;

	start = MPI_Wtime();
	for (int i = 0; i < iterations; i++) {
		computations();
	}
	end = MPI_Wtime();
	double elapsed_no_barrier = end - start;
	
	if (world_rank == 0)
		fprintf(stderr, "Barrier: %lf [s], no barrier: %lf [s]\n", elapsed_barrier, elapsed_no_barrier);

	MPI_Finalize();
	return 0;
}

File: lab2/python/barrier.py
#!/usr/bin/python

from math import tan, exp, log

from mpi4py import MPI


comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

number_of_messages = 2000


def computations():
    x = 123456789
    for i in range(100):
        x = x * x % 23456789


comm.Barrier()

start_time = MPI.Wtime()

for _ in xrange(0, number_of_messages):
    comm.Barrier()
    computations()

end_time = MPI.Wtime()

elapsed_barrier = end_time - start_time

start_time = MPI.Wtime()

for _ in xrange(0, number_of_messages):
    computations()

end_time = MPI.Wtime()

elapsed_no_barrier = end_time - start_time

if rank == 0:
    print "Barrier: {0} [s], no barrier: {1} [s]".format(elapsed_barrier, elapsed_no_barrier)
